Unconditionally register Python decomps to Meta key in Python Dispatcher

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85750

This makes them available for Python Dispatcher to service them when
symbolic shapes are involved.  This is needed because under certain
conditions, functionalization will directly call the Meta kernel for a
function in order to produce a properly sized output wrapper tensor
for a view operation. This direct call bypasses the normal decomposition
table mechanism.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Unconditionally register Python decomps to Meta key in Python Dispatcher

This makes them available for Python Dispatcher to service them when
symbolic shapes are involved.  I am not exactly sure why this is needed
but empirically pytorch_BERT doesn't work without it, and this change
seems morally right.  I'll try to add a test when I figure it out.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
the failing tests here pass on the branch, so I need more stuff...

eellison commented:
Do you mind if i land https://github.com/pytorch/pytorch/pull/85471 first ?

ezyang commented:
Pleae do. These look orthogonal, and this PR is failing tests anyway

wconstab reviewed (approved):
I think this makes sense.  Python dispatcher would only pick this meta kernel if there was a matching DK for meta, IIUC.

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -100,6 +100,7 @@ def add_op_to_table(aten_op):
                 if op_overload in registry:
                     raise RuntimeError(f"duplicate registrations for {op_overload}")
                 registry[op_overload] = fn
+                op_overload.py_impl(torch._C.DispatchKey.Meta)(fn)
```
Hmmm won't this unconditionally register a bunch of our python decomps to the meta key, and stomp over some of our C++ decomps/meta kernels?

bdhirsh replied:
oh nvm is this fine because the PythonDispatcher needs to be explicitly turned on elsewhere, anyway.

ezyang committed:
Update on "Unconditionally register Python decomps to Meta key in Python Dispatcher"


This makes them available for Python Dispatcher to service them when
symbolic shapes are involved.  This is needed because under certain
conditions, functionalization will directly call the Meta kernel for a
function in order to produce a properly sized output wrapper tensor
for a view operation. This direct call bypasses the normal decomposition
table mechanism.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
